### ISSUE TITLE
__set_stack_limits not found in SIDE_MODULE=2

### ISSUE THREAD

**Issue by Mintyboi** [emscripten-core/emscripten#13709]
`__set_stack_limits` gets optimized away when I build a side module via `-s SIDE_MODULE=2`.
Then when I try to load this side module into the main module (built with `-s MAIN_MODULE=2 -s ASSERTIONS=2`), I get an error:

```
TypeError: moduleExports.__set_stack_limits is not a function
``` 
due to the STACK_OVERFLOW_CHECK when `ASSERTIONS=2`

**Comment by Mintyboi** [comment]
A little more info. Even after building side module with `-s SIDE_MODULE=1`, the function is a no-op.
```
(func $__set_stack_limits (type $t1) (param $p0 i32) (param $p1 i32)
    nop)
```
Do we need this check on side modules?

**Comment by sbc100** [comment]
Ah I see, because the stack start/end are never actually accessed in the side module binaryen removes the globals that represent them.

If you do something that uses the shadow stack in your side module then at least the globals should not be removed. e.g.:

```
$ cat side.c
#include <emscripten.h>
#include <stdlib.h>

void bar(void*);

EMSCRIPTEN_KEEPALIVE int foo() {
  void* x = alloca(1024);
  bar(x);
  return 0;
}
```

Perhaps we should make the `__set_stack_limits` export optional, or prevent its removal (whereever that is happening).

**Comment by Mintyboi** [comment]
Right, I get the `__set_stack_limits` function after using your code and allocating the stack memory.

How do we proceed from here? Seeing that `__set_stack_limits` does not make sense in the side module when we do not access the stack, shall I make the change in postInstantiation() to make sure that `__set_stack_limits` exists before calling it?

**Comment by sbc100** [comment]
I'm not sure which of the two options is better:

1. Make the `__set_stack_limits` export optional
2. Prevent DCE of `__set_stack_limits`

(2) seems like it might better because:

a.  `__set_stack_limits` is only used when `STACK_OVERFLOW_CHECK >= 2` so actually doing DCE to reduce code size is not imported in that configuration.
b. It means that if `__set_stack_limits` is not created by binaryen for whatever reason we get an error rather then silently not working.

@kripken WDYT?

**Comment by sbc100** [comment]
BTW I think the reason that `-s EXPORTED_FUNCTIONS=['__set_stack_limits']` didn't work is that you are missing the leading `_`.   The EXPORTED_FUNCTIONS list needs to have `_` prefix on all native functions so that means `__set_stack_limits` needs three leading underscores :(

Can you verify that works as a workaround for now?

**Comment by kripken** [comment]
Preventing DCE sounds right to me as well. That is, in the mode where we use that function, we should add it to `EXPORTED_FUNCTIONS` so that it is preserved like all other necessary things.

**Comment by Mintyboi** [comment]
>Can you verify that works as a workaround for now?

Thanks for the suggestion @sbc100, unfortunately that doesn't work either. I'll try dig in more to see in which pass of binaryen, this gets deleted.

**Comment by sbc100** [comment]
Strange.. I'm pretty sure `-s EXPORTED_FUNCTIONS=['__set_stack_limits']` (with 3 underscores) makes the issue go away for me.

**Comment by Mintyboi** [comment]
This is the code I used
```
#include <emscripten.h>
#include <stdlib.h>

void bar(void*);

EMSCRIPTEN_KEEPALIVE int foo() {
  bar((void*)1);
  return 0;
}
```

and the build flags
`emcc side.cpp -Oz -s SIDE_MODULE=2 -s ASSERTIONS=2 -s EXPORTED_FUNCTIONS=['___set_stack_limits'] -o ./side.wasm`

and the text output for side.wasm
```
(module
  (type $t0 (func))
  (type $t1 (func (param i32)))
  (type $t2 (func (result i32)))
  (import "env" "_Z3barPv" (func $env._Z3barPv (type $t1)))
  (import "env" "__memory_base" (global $env.__memory_base i32))
  (import "env" "__table_base" (global $env.__table_base i32))
  (import "env" "memory" (memory $env.memory 0))
  (import "env" "__indirect_function_table" (table $env.__indirect_function_table 0 funcref))
  (func $__post_instantiate (type $t0)
    call $f2)
  (func $f2 (type $t0))
  (func $_Z3foov (type $t2) (result i32)
    (local $l0 i32) (local $l1 i32)
    i32.const 1
    local.set $l0
    local.get $l0
    call $env._Z3barPv
    i32.const 0
    local.set $l1
    local.get $l1
    return)
  (export "__post_instantiate" (func $__post_instantiate))
  (export "_Z3foov" (func $_Z3foov)))
```
